core/chains/evm/gas/universal_estimator.go

```diff
@@ -0,0 +1,413 @@
+		Help: "Sets latest BaseFee (in Wei)",
+	},
+		[]string{"evmChainID"},
+	)
+	promUniversalEstimatorMaxPriorityFeePerGas = promauto.NewGaugeVec(prometheus.GaugeOpts{
+		Name: "max_priority_fee_per_gas_updater",
+		Help: "Sets latest MaxPriorityFeePerGas (in Wei)",
+	},
+		[]string{"evmChainID"},
+	)
+	promUniversalEstimatorMaxFeePerGas = promauto.NewGaugeVec(prometheus.GaugeOpts{
+		Name: "max_fee_per_gas_updater",
+		Help: "Sets latest MaxFeePerGas (in Wei)",
+	},
+		[]string{"evmChainID"},
+	)
+)
+
+const (
+	MinimumBumpPercentage   = 10 // based on geth's spec
+	ConnectivityPercentile  = 85
+	BaseFeeBufferPercentage = 40
+)
+
+type UniversalEstimatorConfig struct {
+	BumpPercent  uint16
+	CacheTimeout time.Duration
+	EIP1559      bool
+
+	BlockHistorySize uint64
+	RewardPercentile float64
+	HasMempool       bool
+}
+
+type universalEstimatorClient interface {
+	SuggestGasPrice(ctx context.Context) (*big.Int, error)
+	FeeHistory(ctx context.Context, blockCount uint64, rewardPercentiles []float64) (feeHistory *ethereum.FeeHistory, err error)
+}
+
+type UniversalEstimator struct {
+	services.StateMachine
+
+	client  universalEstimatorClient
+	logger  logger.Logger
+	config  UniversalEstimatorConfig
+	chainID *big.Int
+
+	gasPriceMu sync.RWMutex
+	gasPrice   *assets.Wei
+
+	dynamicPriceMu sync.RWMutex
+	dynamicPrice   DynamicFee
+
+	priorityFeeThresholdMu sync.RWMutex
+	priorityFeeThreshold   *assets.Wei
+
+	l1Oracle rollups.L1Oracle
+
+	wg     *sync.WaitGroup
+	stopCh services.StopChan
+}
+
+func NewUniversalEstimator(lggr logger.Logger, client universalEstimatorClient, cfg UniversalEstimatorConfig, chainID *big.Int, l1Oracle rollups.L1Oracle) *UniversalEstimator {
+	return &UniversalEstimator{
+		client:   client,
+		logger:   logger.Named(lggr, "UniversalEstimator"),
+		config:   cfg,
+		chainID:  chainID,
+		l1Oracle: l1Oracle,
+		wg:       new(sync.WaitGroup),
+		stopCh:   make(chan struct{}),
+	}
+}
+
+func (u *UniversalEstimator) Start(context.Context) error {
+	return u.StartOnce("UniversalEstimator", func() error {
+		if u.config.BumpPercent < MinimumBumpPercentage {
+			return fmt.Errorf("BumpPercent: %s is less than minimum allowed percentage: %s",
+				strconv.FormatUint(uint64(u.config.BumpPercent), 10), strconv.Itoa(MinimumBumpPercentage))
+		}
+		if u.config.EIP1559 && u.config.RewardPercentile > ConnectivityPercentile {
+			return fmt.Errorf("RewardPercentile: %s is greater than maximum allowed percentile: %s",
+				strconv.FormatUint(uint64(u.config.RewardPercentile), 10), strconv.Itoa(ConnectivityPercentile))
+		}
+		if u.config.EIP1559 && u.config.BlockHistorySize == 0 {
+			return fmt.Errorf("BlockHistorySize is set to 0 and EIP1559 is enabled")
+		}
+		u.wg.Add(1)
+		go u.run()
+
+		return nil
+	})
+}
+
+func (u *UniversalEstimator) Close() error {
+	return u.StopOnce("UniversalEstimator", func() error {
+		close(u.stopCh)
+		u.wg.Wait()
+		return nil
+	})
+}
+
+func (u *UniversalEstimator) run() {
```

Comment:
It's looks like there's some documentation about how the UE works in the FetchDynamicPrice() function, I think I think it'd be good to give a high level overview of how the whole thing works here.